src/message-stream.ts

```diff
@@ -268,65 +321,90 @@ export class MessageStream extends PassThrough {
     client.initialize();
     return client.subscriberStub as Promise<ClientStub>;
   }
+
   /**
    * Since we do not use the streams to ack/modAck messages, they will close
    * by themselves unless we periodically send empty messages.
    *
    * @private
    */
   private _keepAlive(): void {
-    this._streams.forEach((receivedStatus, stream) => {
-      // its possible that a status event fires off (signaling the rpc being
-      // closed) but the stream hasn't drained yet, writing to this stream will
-      // result in a `write after end` error
-      if (!receivedStatus) {
-        stream.write({});
+    this._streams.forEach(tracker => {
+      // It's possible that a status event fires off (signaling the rpc being
+      // closed) but the stream hasn't drained yet. Writing to such a stream will
+      // result in a `write after end` error.
+      if (!tracker.receivedStatus && tracker.stream) {
+        tracker.stream.write({});
       }
     });
   }
+
+  // Returns the number of tracked streams that contain an actual stream (good or not).
+  private _activeStreams(): number {
+    return this._streams.reduce((p, t) => (t.stream ? 1 : 0) + p, 0);
+  }
+
   /**
    * Once the stream has nothing left to read, we'll remove it and attempt to
    * refill our stream pool if needed.
    *
    * @private
    *
-   * @param {Duplex} stream The ended stream.
+   * @param {number} index The ended stream.
    * @param {object} status The stream status.
    */
-  private _onEnd(stream: PullStream, status: grpc.StatusObject): void {
-    this._removeStream(stream);
-
-    if (this._fillHandle) {
-      return;
-    }
-
-    if (this._retrier.retry(status)) {
-      const delay = this._retrier.createTimeout();
-      this._fillHandle = setTimeout(() => this._fillStreamPool(), delay);
-    } else if (!this._streams.size) {
+  private _onEnd(index: number, status: grpc.StatusObject): void {
+    this._removeStream(index);
+
+    if (PullRetry.retry(status)) {
+      this.emit(
+        'debug',
+        new Error(
```

Comment:
I admit it's a bit of an abuse of what Error is supposed to do, but basically when you `new Error`, it's just making an object with some messages and a stack trace. There shouldn't be any other implications there. If that turns out to be wrong (I don't think it is though) we can always patch these later since they're not "load bearing" API surface. (Just advisory things the user might log, for the most part.)

Reply:
I went ahead and changed this around to be more explicitly advisory/loggy.